MesaModel/run.py

```diff
@@ -69,66 +77,81 @@ def run_model(
-                    "pupil_inattentiveness": UserSettableParameter(
-                        "slider", "Use Pupil Inattentiveness ", 1.0, 0.00, 1.0, 1.0
-                    ),
-                    "pupil_hyper_impulsivity": UserSettableParameter(
-                        "slider", "Use Pupil Hyperactivity ", 1.0, 0.00, 1.0, 1.0
-                    ),
-                    "pupil_attention_span": UserSettableParameter(
-                        "slider", "Pupil Attention Span Limit", 5.0, 0.00, 5.0, 1.0
-                    ),
-                    "write_file": True,
-                },
-            )
-
-            server.launch()
-        else:
-            teacher_params = TeacherParamType(1, 1)
-            pupil_params = PupilParamType(0, 0, 2)
-
-            model = SimModel(
-                all_data,
-                model_initial_state,
-                output_data,
-                fixed_params=(class_id, teacher_params, pupil_params),
-            )
-            model.run_model()
-
-    output_data.write_file()
+    output_data_writer = OutputDataWriter(output_file)
+
+    # Get data first to determine grid size
+    model_initial_state = ModelState(0, 0, 0, 0, 0)
+    canvas_grid = create_canvas_grid(GridParamType(8, 8))
+    click.echo(f"Running on class {class_ids}")
+
+    if webserver:
+        server = ModularServer(
+            SimModel,
+            [canvas_grid, sim_element, sim_chart],
+            "Classroom ABM",
+            {
+                "all_data": all_data,
+                "model_initial_state": model_initial_state,
+                "output_data": output_data_writer,
+                "canvas_grid": canvas_grid,
+                "instructions": UserSettableParameter(
+                    "static_text",
+                    value="Modify the parameters below then click Reset to update the model.",
+                ),
+                "class_id": UserSettableParameter(
+                    "choice", "Class ID", value=class_ids[0], choices=class_ids
+                ),
+                "teacher_quality": UserSettableParameter(
+                    "slider", "Teaching quality", 5.0, 0.00, 5.0, 1.0
+                ),
+                "teacher_control": UserSettableParameter(
+                    "slider", "Teaching control", 5.0, 0.00, 5.0, 1.0
+                ),
+                "pupil_inattentiveness": UserSettableParameter(
+                    "slider", "Use Pupil Inattentiveness ", 1.0, 0.00, 1.0, 1.0
+                ),
+                "pupil_hyper_impulsivity": UserSettableParameter(
+                    "slider", "Use Pupil Hyperactivity ", 1.0, 0.00, 1.0, 1.0
+                ),
+                "pupil_attention_span": UserSettableParameter(
+                    "slider", "Pupil Attention Span Limit", 5.0, 0.00, 5.0, 1.0
+                ),
+            },
+        )
+
+        server.launch()
+
+    else:
+        # Num processes is set to 2 for now. If we don't pass the nr_processes
```

Comment:
Could we pass num_processes on the command line, defaulting to 2?

Reply:
Good idea. Have updated the readme with the new options too.